radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java

```diff
@@ -0,0 +1,191 @@
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.consensus;
+
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anySet;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.ImmutableList;
+import com.radixdlt.common.Atom;
+import com.radixdlt.consensus.ChainedBFT.Event;
+import com.radixdlt.consensus.liveness.PacemakerImpl;
+import com.radixdlt.consensus.liveness.ProposalGenerator;
+import com.radixdlt.consensus.liveness.ProposerElection;
+import com.radixdlt.consensus.liveness.RotatingLeaders;
+import com.radixdlt.consensus.liveness.Dictatorship;
+import com.radixdlt.consensus.safety.QuorumRequirements;
+import com.radixdlt.consensus.safety.SafetyRules;
+import com.radixdlt.consensus.safety.SafetyState;
+import com.radixdlt.consensus.safety.WhitelistQuorum;
+import com.radixdlt.crypto.ECDSASignatures;
+import com.radixdlt.crypto.ECKeyPair;
+import com.radixdlt.engine.RadixEngine;
+import com.radixdlt.mempool.Mempool;
+import com.radixdlt.network.TestEventCoordinatorNetwork;
+import io.reactivex.rxjava3.core.Observable;
+import io.reactivex.rxjava3.observers.TestObserver;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import org.apache.commons.math3.util.Pair;
+import org.junit.Before;
+import org.junit.Test;
+import org.radix.logging.Logger;
+import org.radix.logging.Logging;
+
+public class MultiNodeConsensusTest {
+	private static final Logger log = Logging.getLogger("Test");
+	private final TestEventCoordinatorNetwork testEventCoordinatorNetwork = new TestEventCoordinatorNetwork();
+
+	private Atom genesis;
+	private Vertex genesisVertex;
+	private QuorumCertificate genesisQC;
+
+	@Before
+	public void setup() {
+		this.genesis = mock(Atom.class);
+		when(this.genesis.toString()).thenReturn(Long.toHexString(0));
+		this.genesisVertex = Vertex.createGenesis(genesis);
+		this.genesisQC = new QuorumCertificate(
+			new VertexMetadata(View.genesis(), genesisVertex.getId(), null, null),
+			new ECDSASignatures()
+		);
+	}
+
+	private ChainedBFT createBFTInstance(
+		ECKeyPair key,
+		ProposerElection proposerElection,
+		QuorumRequirements quorumRequirements,
+		VertexStore vertexStore
+	) {
+		Mempool mempool = mock(Mempool.class);
```

Comment:
Why are we mocking mempool but nothing else?

Reply:
mempool is the only module that sort exists outside of BFT consensus space, so can mock it so we don't need to worry about client interactions with system.